php_phongo.c

```diff
@@ -1921,6 +1921,17 @@ static bool php_phongo_apply_wc_options_to_uri(mongoc_uri_t* uri, bson_t* option
 } /* }}} */
 
 #ifdef MONGOC_ENABLE_SSL
+
+/* This is copied from mongoc-ssl.c */
+static void php_phongo_mongoc_ssl_opts_from_uri(mongoc_ssl_opt_t* ssl_opt, mongoc_uri_t* uri)
+{
+	ssl_opt->pem_file               = mongoc_uri_get_option_as_utf8(uri, MONGOC_URI_SSLCLIENTCERTIFICATEKEYFILE, NULL);
+	ssl_opt->pem_pwd                = mongoc_uri_get_option_as_utf8(uri, MONGOC_URI_SSLCLIENTCERTIFICATEKEYPASSWORD, NULL);
+	ssl_opt->ca_file                = mongoc_uri_get_option_as_utf8(uri, MONGOC_URI_SSLCERTIFICATEAUTHORITYFILE, NULL);
```

Comment:
[`mongoc_uri_get_option_as_utf8()`](http://mongoc.org/libmongoc/current/mongoc_uri_get_option_as_utf8.html) returns a `const char *` that is only valid for the lifetime of the URI structure.

I think these need to be run through `estrdup()` as we do in [`php_phongo_fetch_ssl_opt_string()`](https://github.com/mongodb/mongo-php-driver/blob/1.5.1/php_phongo.c#L1924) (if the string from `php_array.h` is not already duplicated). Ideally, we'd use `estrndup()` here but we don't have the string length (and real values likely shouldn't include null bytes).

Before you fix this, are you able to observe a double-free with this current implementation when `php_phongo_free_ssl_opt()` is called and runs `str_efree` on these three string fields?

Reply:
Note that you'll probably want to create a local `const char *` to re-use for each of these three options:

 * Declare `const char * tmp`
 * `tmp = mongoc_uri_get_option_as_utf8(...)`
 * If `tmp` isn't null, `strdup()` and assign to the struct

Reply:
At the end of manager initialization, the `mongoc_ssl_opt_t` returned by `php_phongo_make_ssl_opt()` [will be freed](https://github.com/mongodb/mongo-php-driver/blob/1.5.1/php_phongo.c#L2462) via [`php_phongo_free_ssl_opt()`](https://github.com/mongodb/mongo-php-driver/blob/1.5.1/php_phongo.c#L2009). That will involve calling `str_efree()` on every non-NULL string on the struct before we `efree()` the struct pointer itself.

Currently, some of those strings may end up being `const char *` that point to memory within the `mongoc_uri_t`. That data will have been freed when [`phongo_manager_init()` calls [`mongoc_uri_destroy()`](https://github.com/mongodb/mongo-php-driver/blob/1.5.1/php_phongo.c#L2458) immediately before the `php_phongo_free_ssl_opt()` call. This can cause the double-free.

When `mongoc_client_set_ssl_opts()` talks about duplicating strings, it's referring to copying the `mongoc_ssl_opt_t` into the client/pool. In addition to copying the struct, it duplicates any non-NULL strings to avoid referring back to any memory from the original `mongoc_ssl_opt_t` that was copied. In libmongoc's case, they [use a stack-allocated struct](https://github.com/mongodb/mongo-c-driver/blob/1.12.0/src/libmongoc/src/mongoc/mongoc-client.c#L1030) to quickly collect strings from the URI and ferry them over to a client/pool. This stack-allocated `mongoc_ssl_opt_t` only exists for a subset of the corresponding `mongoc_uri_t` lifetime, so they don't have to worry about freeing pointers like `mongoc_ssl_opt_t.pem_pwd` at all.